File: crates/tacet-napi/examples/simple.ts
/**
 * Simple example: Testing crypto operations for timing leaks.
 *
 * Run with: bun examples/simple.ts
 */

import { TimingOracle, AttackerModel, calibrateTimer } from "../dist/index.js";
import crypto from "node:crypto";

// Show timer info
const timerInfo = calibrateTimer();
console.log("Timer info:");
console.log(`  Resolution: ${timerInfo.resolutionNs.toFixed(1)}ns`);
console.log(`  Frequency: ${(timerInfo.frequencyHz / 1e9).toFixed(2)} GHz\n`);

// Test 1: SHA-256 hashing (should PASS - constant-time)
console.log("Test 1: SHA-256 hashing (should PASS)...\n");

const result1 = TimingOracle.forAttacker(AttackerModel.AdjacentNetwork)
  .timeBudget(15_000) // 15 seconds
  .maxSamples(50_000)
  .showProgress()
  .test(
    {
      baseline: () => Buffer.alloc(64, 0), // All zeros
      sample: () => crypto.randomBytes(64), // Random data
    },
    (input) => {
      // SHA-256 is constant-time
      crypto.createHash("sha256").update(input).digest();
    }
  );

console.log("\nResult:", result1.outcomeString());
console.log("Leak probability:", result1.leakProbabilityPercent());
console.log("Batching:", result1.batchingInfo.rationale);
console.log(`Effect: max=${result1.effect.maxEffectNs.toFixed(2)}ns`);

if (result1.isPass()) {
  console.log("PASS: No timing leak detected\n");
} else if (result1.isFail()) {
  console.log("FAIL: Timing leak detected!");
  console.log("Exploitability:", result1.exploitabilityString());
} else if (result1.isInconclusive()) {
  console.log("INCONCLUSIVE:", result1.inconclusiveReasonString(), "\n");
} else {
  console.log("UNMEASURABLE:", result1.recommendation, "\n");
}

// Test 2: Early-exit comparison (should FAIL - timing leak)
console.log("---\nTest 2: Early-exit comparison (should FAIL)...\n");

const secret = Buffer.from("supersecretkey!!");

const result2 = TimingOracle.forAttacker(AttackerModel.AdjacentNetwork)
  .timeBudget(15_000)
  .maxSamples(50_000)
  .showProgress()
  .test(
    {
      baseline: () => Buffer.from(secret), // Matches secret (fast path)
      sample: () => crypto.randomBytes(16), // Usually differs early (slow path)
    },
    (input) => {
      // INSECURE: Early-exit comparison (has timing leak!)
      for (let i = 0; i < input.length; i++) {
        if (input[i] !== secret[i]) {
          return; // Early exit on mismatch - TIMING LEAK
        }
      }
    }
  );

console.log("\nResult:", result2.outcomeString());
console.log("Leak probability:", result2.leakProbabilityPercent());
console.log("Batching:", result2.batchingInfo.rationale);
console.log(`Effect: max=${result2.effect.maxEffectNs.toFixed(2)}ns`);

if (result2.isPass()) {
  console.log("PASS: No timing leak detected (unexpected!)\n");
} else if (result2.isFail()) {
  console.log("FAIL: Timing leak detected! (expected)");
  console.log("Exploitability:", result2.exploitabilityString());
} else if (result2.isInconclusive()) {
  console.log("INCONCLUSIVE:", result2.inconclusiveReasonString(), "\n");
} else {
  console.log("UNMEASURABLE:", result2.recommendation, "\n");
}

// Demonstrate assertion style
console.log("\n---\nTest 3: Assertion style (assertNoLeak)...\n");

try {
  result2.assertNoLeak();
  console.log("No leak detected");
} catch (e) {
  console.log("Caught TimingLeakError:", (e as Error).message);
}

File: crates/tacet/examples/simple.rs
//! Simple example demonstrating basic tacet usage.
//!
//! This demonstrates the CORRECT way to test operations:
//! - Pre-generate all inputs before measurement
//! - The test closure executes identical code for fixed and random inputs
//! - Only the input data differs

use std::time::Duration;
use tacet::{helpers::InputPair, timing_test_checked, AttackerModel, Outcome, TimingOracle};

fn main() {
    println!("tacet simple example\n");

    // Example: Testing a potentially leaky comparison
    let secret = [0u8; 32];

    // Pre-generate inputs using InputPair
    let inputs = InputPair::new(
        || [0u8; 32], // Baseline: all zeros (same as secret)
        || {
            let mut arr = [0u8; 32];
            for item in &mut arr {
                *item = rand::random();
            }
            arr
        },
    );

    // Simple API with attacker model
    let outcome = TimingOracle::for_attacker(AttackerModel::AdjacentNetwork)
        .time_budget(Duration::from_secs(30))
        .test(inputs, |data| {
            compare_bytes(&secret, data);
        });

    match &outcome {
        Outcome::Pass {
            leak_probability,
            quality,
            diagnostics,
            ..
        } => {
            println!("Result: PASS");
            println!("Leak probability: {:.1}%", leak_probability * 100.0);
            println!("Quality: {:?}", quality);
            println!("Timer resolution: {:.1}ns", diagnostics.timer_resolution_ns);
        }
        Outcome::Fail {
            leak_probability,
            exploitability,
            effect,
            ..
        } => {
            println!("Result: FAIL");
            println!("Leak probability: {:.1}%", leak_probability * 100.0);
            println!("Exploitability: {:?}", exploitability);
            println!(
                "Max effect: {:.1}ns (95% CI: {:.1}–{:.1}ns)",
                effect.max_effect_ns, effect.credible_interval_ns.0, effect.credible_interval_ns.1
            );
        }
        Outcome::Inconclusive {
            reason,
            leak_probability,
            ..
        } => {
            println!("Result: INCONCLUSIVE");
            println!("Leak probability: {:.1}%", leak_probability * 100.0);
            println!("Reason: {:?}", reason);
        }
        Outcome::Unmeasurable { recommendation, .. } => {
            println!("Could not measure: {}", recommendation);
            return;
        }
        Outcome::Research(research) => {
            println!("Result: RESEARCH MODE");
            println!("Max effect: {:.1}ns", research.max_effect_ns);
            println!("Status: {:?}", research.status);
            return;
        }
    }

    // Using timing_test_checked! macro with custom config
    let outcome = timing_test_checked! {
        oracle: TimingOracle::for_attacker(AttackerModel::AdjacentNetwork)
            .time_budget(Duration::from_secs(10)),
        baseline: || [0u8; 32],
        sample: || {
            let mut arr = [0u8; 32];
            for item in &mut arr {
                *item = rand::random();
            }
            arr
        },
        measure: |data| {
            compare_bytes(&secret, data);
        },
    };

    if let Outcome::Pass {
        leak_probability, ..
    }
    | Outcome::Fail {
        leak_probability, ..
    } = outcome
    {
        println!("\nWith custom config:");
        println!("Leak probability: {:.1}%", leak_probability * 100.0);
    }
}

/// Non-constant-time comparison (intentionally leaky for demo).
fn compare_bytes(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    for i in 0..a.len() {
        if a[i] != b[i] {
            return false; // Early exit - timing leak!
        }
    }
    true
}
